### PR TITLE
User timer based rendering instead of vsync

### LINKED ISSUE #824: Redraw randomly hangs until a keypress
Using Alacritty (commit 6916537858d12e02478d618339d1b9f0e73a677a) on Ubuntu 16.04 with X11.

It is better described with the following screencast. Here I'm repeatedly opening and closing `tig`, so the output should be the same in each iteration. The screen often stays in a partially redrawed state (marked by mouse cursor movement). Then it finishes redraw correctly when I'm pressing a key.

screencast

### PR DESCRIPTION
**Pull request by kchibisov**
Fixes #824.

--
For me it was keeping +- target frame rate with multiple instances. Would suggest to test with multiple monitors **at different refresh rate**, since I can't test it myself...

### CODE REVIEW

File: alacritty/src/display/mod.rs [hunk 1]
```diff
@@ -487,14 +492,8 @@ impl Display {
...
-        // We use vsync everywhere except wayland.
-        if !is_wayland {
-            if let Err(err) =
-                surface.set_swap_interval(&context, SwapInterval::Wait(NonZeroU32::new(1).unwrap()))
-            {
-                warn!("Error setting vsync: {:?}", err);
-            }
-        }
+        // Disable vsync.
+        let _ = surface.set_swap_interval(&context, SwapInterval::DontWait);
```
Comment (kchibisov) [inline alacritty/src/display/mod.rs:496]: I'm not sure how to handle failure of this. It could fail just because there's no extension. Maybe we should distinguish the failure from no extension and from the actual changing?

File: alacritty/src/display/mod.rs [hunk 2]
```diff
@@ -367,6 +368,9 @@ pub struct Display {
     /// The ime on the given display.
     pub ime: Ime,
 
+    /// The state of the timer based for frame scheduling.
+    pub frame_timer_info: FrameTimerInfo,
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:372]:
```suggestion
    /// The state of the timer for frame scheduling.
    pub frame_timer_info: FrameTimerInfo,
```

File: alacritty/src/display/mod.rs (hunk 1 above)
Comment (chrisduerr) [reply alacritty/src/display/mod.rs:495]: I'm fine with ignoring it. We could log it maybe?

File: alacritty/src/display/mod.rs [hunk 3]
```diff
@@ -969,7 +969,9 @@ impl Display {
         // Frame event should be requested before swaping buffers, since it requires surface
         // `commit`, which is done by swap buffers under the hood.
         #[cfg(all(feature = "wayland", not(any(target_os = "macos", windows))))]
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:971]: I think this `cfg` can be removed? Sure it's a runtime check more if Wayland is disabled but it should basically be free and the function exists everywhere now.

Otherwise I'd suggest adding cfgs to all the `!self.is_wayland` places too, we should just be consistent.

File: alacritty/src/display/mod.rs [hunk 4]
```diff
@@ -969,7 +969,9 @@ impl Display {
         // Frame event should be requested before swaping buffers, since it requires surface
         // `commit`, which is done by swap buffers under the hood.
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:970]:
```suggestion
        // Frame event should be requested before swapping buffers on Wayland, since it requires surface
        // `commit`, which is done by swap buffers under the hood.
```

File: alacritty/src/display/mod.rs [hunk 5]
```diff
@@ -1371,23 +1377,38 @@ impl Display {
...
-        should_draw.store(false, Ordering::Relaxed);
+            return;
+        }
 
-        // Request a new frame.
-        surface.frame().quick_assign(move |_, _, _| {
-            should_draw.store(true, Ordering::Relaxed);
-        });
+        // Get the display vblank interval in micro seconds.
+        let monitor_vblank_interval = Duration::from_micros(
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1396]: Why are you using microseconds here? I find that extremely strange considering basically everyone uses either ms or nanos usually.

Just for the sub-millis accuracy without having to use nanos which would be overly precise?

File: alacritty/src/display/mod.rs [hunk 6]
```diff
@@ -1371,23 +1377,38 @@ impl Display {
...
+            (1e6 / (self
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1397]: I find the `1e6` notation difficult to grasp at a glance and I rarely see that in source code likely for that reason. Sure it's shorter, but is it more readable than `1_000_000`?

If you go into the billions and higher it might be more helpful, but I think people will recognize much faster that `1_000_000` is a million rather than `1e6`.

This also applies to the `1e3` used later which is even worse since it's just 1000.

File: alacritty/src/display/mod.rs [hunk 7]
```diff
@@ -1371,23 +1377,38 @@ impl Display {
...
+                .window
+                .current_monitor()
+                .and_then(|monitor| monitor.refresh_rate_millihertz())
+                .unwrap_or(60_000) as f64
+                / 1e3)) as u64,
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1402]: What's the point of this division? Wouldn't it be easier to just do `1_000_000_000 / 60_000` rather than `1_000_000 / (60_000 / 1000)`?

File: alacritty/src/display/mod.rs [hunk 8]
```diff
@@ -1371,23 +1377,38 @@ impl Display {
...
+        );
+
+        let swap_delay = self.frame_timer_info.compute_delay(monitor_vblank_interval);
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1405]: Not sure `delay` is the right word here? Maybe `timeout` is more appropriate? Though I don't love it either.

File: alacritty/src/display/mod.rs [hunk 9]
```diff
@@ -1529,6 +1550,54 @@ impl<T> DerefMut for Replaceable<T> {
     }
 }
 
+/// The frame timer state.
+#[derive(Default)]
+pub struct FrameTimerInfo {
+    /// The optimal delay for the renderer.
+    swap_delay: Option<Duration>,
+
+    /// The time stamp of the previous contiguos frame.
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1559]:
```suggestion
    /// The timestamp of the previous contiguous frame.
```

Besides the typo, what is "contiguous frame" supposed to mean here? Isn't it already implied by "previous" that we're talking about the frame right before this?

I think just this should be better:

```suggestion
    /// The timestamp of the previous frame.
```

Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1559]:
```suggestion
    /// The timestamp of the previous contiguous frame.
```

Besides the typo, what is "contiguous frame" supposed to mean here? Isn't it already implied by "previous" that we're talking about the frame right before this?

I think just this should be better:

```suggestion
    /// The timestamp of the previous frame.
```

File: alacritty/src/display/mod.rs [hunk 10]
```diff
@@ -1529,6 +1550,54 @@ impl<T> DerefMut for Replaceable<T> {
...
+        // If the refresh rate has changed, reset the timings.
+        if self.refresh_interval != refresh_interval || self.swap_delay.is_none() {
+            self.refresh_interval = refresh_interval;
+            self.frame_timestamp = Some(Instant::now());
+            self.swap_delay = Some(refresh_interval);
+            return refresh_interval;
+        }
+
+        // At this point we know that `swap_delay` has some value.
+        let swap_delay = self.swap_delay.unwrap();
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1579]: Why not use this:

```rust
let swap_delay = match self.sway_delay.filter(|_| self.refresh_interval == refresh_interval) {
    Some(swap_delay) => swap_delay,
    None => {
        // Reset stuff
    },
};
```

Maybe a little worse for performance, but should basically be irrelevant.

File: alacritty/src/display/mod.rs [hunk 11]
```diff
@@ -1529,6 +1550,54 @@ impl<T> DerefMut for Replaceable<T> {
...
+
+        let new_swap_delay = match self.frame_timestamp {
+            Some(last_frame_timestamp) => {
+                // Compute the error.
+                let error = self.refresh_interval.as_micros() as i64
+                    - last_frame_timestamp.elapsed().as_micros() as i64;
+                let new_swap_delay = swap_delay.as_micros() as i64 + error;
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1586]: Assuming that a frame took a particularly long time, that will mean that immediately the next frame will get hurt a lot too. But I don't think there's any immediate connection between the last frame and the next frame, right?

So shouldn't we try to "stabilize" towards some rolling average rather than flip-flopping between a massive and tiny delay?

What you seem to be doing is basically just to try and **start** rendering every *refresh_rate* ms, while realistically what I think we'd want is to try and **finish** rendering every *refresh_rate* ms.

So I feel like we should add a rolling average to this and subtract it as a "buffer" for how long we'd expect rendering to take. Then keep your logic to adjust for when that buffer was insufficient or to much to ensure we're staying on track. But that way we should be able to aim to finish in time rather than starting in time.

File: alacritty/src/event.rs [hunk 12]
```diff
@@ -1474,6 +1474,13 @@ impl Processor {
                             WinitEvent::RedrawEventsCleared,
                         );
                     }
+
+                    // Update the scheduler after we've done everything, so
+                    // the deadline will be more correct.
+                    *control_flow = match scheduler.update() {
+                        Some(instant) => ControlFlow::WaitUntil(instant),
+                        None => ControlFlow::Wait,
+                    };
```
Comment (chrisduerr) [inline alacritty/src/event.rs:1483]: This comment is useful for my review, but "more correct" as a relative term will be completely useless when you don't have the comparison to the old version anymore.

I'd suggest rewording to something like this:

```suggestion
                    // Update the scheduler after event processing, to ensure the event loop deadline is as accurate as possible.
                    *control_flow = match scheduler.update() {
                        Some(instant) => ControlFlow::WaitUntil(instant),
                        None => ControlFlow::Wait,
                    };
```

Comment (chrisduerr) [inline alacritty/src/event.rs:1483]: This comment is useful for my review, but "more correct" as a relative term will be completely useless when you don't have the comparison to the old version anymore.

I'd suggest rewording to something like this:

```suggestion
                    // Update the scheduler after event processing, to ensure the event loop deadline is as accurate as possible.
                    *control_flow = match scheduler.update() {
                        Some(instant) => ControlFlow::WaitUntil(instant),
                        None => ControlFlow::Wait,
                    };
```

Comment (chrisduerr) [inline alacritty/src/event.rs:1483]: This comment is useful for my review, but "more correct" as a relative term will be completely useless when you don't have the comparison to the old version anymore.

I'd suggest rewording to something like this:

```suggestion
                    // Update the scheduler after event processing, to ensure the event loop deadline is as accurate as possible.
                    *control_flow = match scheduler.update() {
                        Some(instant) => ControlFlow::WaitUntil(instant),
                        None => ControlFlow::Wait,
                    };
```

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: alacritty/src/display/mod.rs (hunk 5 above)
Comment (kchibisov) [reply alacritty/src/display/mod.rs:1396]: That's a good question. Simply ns is not enough, so I've picked next... Could use nanos if you feel like it.

Comment (chrisduerr) [reply alacritty/src/display/mod.rs:1396]: Micros should be fine I guess. The explanation makes enough sense where people shouldn't be completely confused reading this.

File: alacritty/src/display/mod.rs (hunk 9 above)
Comment (kchibisov) [reply alacritty/src/display/mod.rs:1559]: aye, it's now a previous frame, it was a bit different before.

File: alacritty/src/display/mod.rs (hunk 11 above)
Comment (kchibisov) [reply alacritty/src/display/mod.rs:1586]: >Assuming that a frame took a particularly long time, that will mean that immediately the next frame will get hurt a lot too.

The frame will fire immediately in this case. Not sure how correct this will be though.

>But I don't think there's any immediate connection between the last frame and the next frame, right?

The connection was initially made for contiguous frames, but it was showing worse perf then what it's right now. The initial idea was to compute the optimal `timeout` for contiguous frames and use it as a `default` when we don't know sufficient information about 'how much time we should timeout to render at the monitor refresh rate'.

>So shouldn't we try to "stabilize" towards some rolling average rather than flip-flopping between a massive and tiny delay?

It was slower, but maybe I was doing it wrong. Basically what I was trying to do is when we hit `draw` and we skip it due to frame not being dirty I was resetting the last timestamp. But I think what we could try to do is when the error gets too big as in it zeros the delay we might want to ignore adjusting the delay in this case. The large error means that we likely wasn't ready to draw.

>What you seem to be doing is basically just to try and start rendering every refresh_rate ms, while realistically what I think we'd want is to try and finish rendering every refresh_rate ms.

I think the latter is what I'm trying to do since I schedule the frame after all the rendering, so I'm accounting not for the start but for the end. The place is after all the rendering operations and I'm scheduling for that. If I were doing the same thing before the `swap_buffers` then yes, it'll be the former.

>So I feel like we should add a rolling average to this and subtract it as a "buffer" for how long we'd expect rendering to take. 

Rendering is already accounted since we do that after `gl::Finish` and `swap_buffers`, it could be wise to call the `gl::Finish` on other platforms though, however it should be the case on other platforms already with the swap buffers.


>Then keep your logic to adjust for when that buffer was insufficient or to much to ensure we're staying on track. But that way we should be able to aim to finish in time rather than starting in time.

We aiming on finish time already

So to sum it up, what could be adjusted is behavior when swap delay get zeroed, since it indeed breaks some scheduling, however when I was trying to do so with the method I've described above I got lower fps then with the current approach, which makes sense, since I was rendering with 0 delay the next frame as well or something like that.

I also used various techniques, like using `Meter` for averaging the error, but not sure whether it got me anywhere, I think I should retry at least, since last time I did that with wrong scheduler logic and haven't bothered to retest once I changed the scheduler timing re-computation.

File: alacritty/src/display/mod.rs [hunk 13]
```diff
@@ -985,6 +985,8 @@ impl Display {
             self.renderer.finish();
         }
 
+        // XXX request for the rendering frame after swap buffers, so the time
+        // for finishing gl operations will be accounted in to the timeout.
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:989]:
```suggestion
        // XXX: Request the new frame after swapping buffers, so the time to finish OpenGL operations is accounted for in the timeout.
```

**Comment by kchibisov** [comment]
@chrisduerr I've updated the logic of rescheduling. It's a bit more stable when running `yes` and also accounts for long delays...

File: alacritty/src/display/mod.rs [hunk 14]
```diff
@@ -367,6 +368,9 @@ pub struct Display {
     /// The ime on the given display.
     pub ime: Ime,
 
+    /// The state of the timer for frame scheduling.
+    pub frame_timer_info: FrameTimerInfo,
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:372]:
```suggestion
    pub frame_timer: FrameTimer,
```

"Info" is a pretty useless suffix usually, I think we can just remove it.

File: alacritty/src/display/mod.rs [hunk 15]
```diff
@@ -1376,23 +1385,40 @@ impl Display {
...
+                .window
+                .current_monitor()
+                .and_then(|monitor| monitor.refresh_rate_millihertz())
+                .unwrap_or(60_000) as f64;
 
-        // Mark that window was drawn.
-        should_draw.store(false, Ordering::Relaxed);
+        // Now convert it to micro seconds.
+        let monitor_vblank_interval =
+            Duration::from_micros((1000. * monitor_vblank_interval) as u64);
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1413]:
```suggestion
            Duration::from_micros(1000 * monitor_vblank_interval as u64);
```

File: alacritty/src/display/mod.rs [hunk 16]
```diff
@@ -1534,6 +1560,63 @@ impl<T> DerefMut for Replaceable<T> {
...
+            self.swap_timeout = refresh_interval;
+            return refresh_interval;
+        }
+
+        // Compute the error.
+        let error = self.refresh_interval.as_micros() as i64
+            - self.frame_timestamp.elapsed().as_micros() as i64;
+
+        // Don't update error and swap timeout if we've waited for too long.
+        if error < self.refresh_interval.as_micros() as i64 {
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1604]: I'm somewhat concerned about this logic here. It will cause an immediate scheduling which I guess is fine but that will mean our "clock" gets reset. Thus the new frame is not in line with our previous frames.

Shouldn't we aim to render at the next "modulo frame interval" time, so that our rendering stays on track and keeps its consistent clock?

Otherwise I'm afraid that if we regularly skip drawing (like every other frame this happens), we'd get very inconsistent latencies (not technically worse, just inconsistent).

File: alacritty/src/display/mod.rs [hunk 17]
```diff
@@ -1534,6 +1560,63 @@ impl<T> DerefMut for Replaceable<T> {
...
+            // Average previously used error with the current one to reduce momentum from the
+            // error sign changes.
+            self.previous_error = (self.previous_error + error) / 2;
+
+            let new_swap_timeout = self.swap_timeout.as_micros() as i64 + self.previous_error;
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1609]: Hm I'm a little concerned with this. Previously since we adjusted for the error **exactly** we'd obviously always draw in `swap_timeout` intervals. But now that we adjust for this error, we have no reference point for our "default" clock cycle anymore. So we'd drift off course since the error correction with the previous frame is not precise.

Instead of taking the `swap_timeout` I feel like we should be calculating the time until the next clock cycle (based on start time) and use that instead?

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: alacritty/src/display/mod.rs (hunk 15 above)
Comment (kchibisov) [reply alacritty/src/display/mod.rs:1413]: No, you'll drop the fractional part?

File: alacritty/src/display/mod.rs (hunk 16 above)
Comment (kchibisov) [reply alacritty/src/display/mod.rs:1604]: But I'm not using `zero`? I'm just using whatever we previously accumulated, just don't recompute and update error?

File: alacritty/src/display/mod.rs (hunk 17 above)
Comment (kchibisov) [reply alacritty/src/display/mod.rs:1609]: >point for our "default" clock cycle anymore. So we'd drift off course since the error correction with the previous frame is not precise.

We use `refresh_interval` for that? So we have a reference, we just try to converge to it slower.

>Instead of taking the swap_timeout I feel like we should be calculating the time until the next clock cycle (based on start time) and use that instead?

Could you elaborate on that? You mean that? For now we use the refresh interval to cap us at refresh rate, do you want some arbitrary logic like get `Instant::now()` at startup, and then draw only on  `start + X * modulo`? I'm not sure it'll be any better to be honest...

File: alacritty/src/display/mod.rs (hunk 15 above)
Comment (chrisduerr) [reply alacritty/src/display/mod.rs:1413]: Ah yeah that's fair, I wasn't considering that we want to keep the fractional precision of the variable.

File: alacritty/src/display/mod.rs (hunk 16 above)
Comment (chrisduerr) [reply alacritty/src/display/mod.rs:1604]: Yeah it's fair that you're not using zero, but you're also not trying to adjust for the clock cycle at all. It doesn't really matter if you're using zero, last `swap_interval` or `swap_timeout` here, none of these will reliably match up with your previous clock cycles.

File: alacritty/src/display/mod.rs (hunk 17 above)
Comment (chrisduerr) [reply alacritty/src/display/mod.rs:1609]: > We use refresh_interval for that? So we have a reference, we just try to converge to it slower.

You have an interval that converges using two historical datapoints. But let's assume this averaging has introduced a 1 microsecond drift in one direction due to rounding. You're never accounting for that drift in the future since all you're doing is just taking the last two errors and averaging them. As a result a drift of 1 microsecond is introduced, shifting your clock permanently.

> Could you elaborate on that? You mean that? For now we use the refresh interval to cap us at refresh rate, do you want some arbitrary logic like get Instant::now() at startup, and then draw only on start + X * modulo? I'm not sure it'll be any better to be honest...

I'm not sure about the exact implementation or your suggestion, but basically yeah. We're already using `elapsed()` to calculate the error, what I'd suggest basically is just doing the same but keeping our clock consistent with the start of the clock rather than consistent with the last frame. That should improve frame consistency.

File: alacritty/src/display/mod.rs [hunk 18]
```diff
@@ -1534,6 +1560,52 @@ impl<T> DerefMut for Replaceable<T> {
...
+        if self.refresh_interval != refresh_interval {
+            self.base = now;
+            self.last_synced_timestamp = now;
+            self.refresh_interval = refresh_interval;
+            return refresh_interval;
+        }
+
+        let next_frame = self.last_synced_timestamp + self.refresh_interval;
+
+        if next_frame < now {
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1596]:
```suggestion
        if next_frame < now {
        // Redraw immediately if we haven't drawn in over `refresh_interval` microseconds.
```

I think adding some comments here would be nice.

File: alacritty/src/display/mod.rs [hunk 19]
```diff
@@ -1534,6 +1560,52 @@ impl<T> DerefMut for Replaceable<T> {
...
+            let elapsed_micros = (now - self.base).as_micros() as u64;
+            let refresh_micros = self.refresh_interval.as_micros() as u64;
+            self.last_synced_timestamp =
+                now - Duration::from_micros(elapsed_micros % refresh_micros);
+            Duration::ZERO
+        } else {
+            self.last_synced_timestamp = next_frame;
```
Comment (chrisduerr) [inline alacritty/src/display/mod.rs:1603]:
```suggestion
            // Redraw on the next `refresh_interval` clock tick.
            self.last_synced_timestamp = next_frame;
```

I think adding some comments here would be nice.

**Review by chrisduerr — APPROVED** [review]
